### PR TITLE
Move extension settings reading into the BaseExtension

### LINKED ISSUE #182: [FEATURE] Move extension settings reading into the BaseExtension
### Is your feature request related to a problem?

#146 added a BaseExtension superclass of Extension implementations to hide a lot of the repetitive implementation details and simplify the actual extensions.

There's some awkward exception handling and initialization in the HelloWorldExtension that could be moved to the BaseExtension.

### What solution would you like?

The `extensionSettings` field and the initialization in the constructor could be moved to the BaseExtension.  
 - The implementing class (HelloWorldExtension) would just pass an `ExtensionSettings` object to `super()`, either:
   - `new ExtensionSettings( ... )`
   - or `Extension.readSettingsFromYaml(EXTENSION_SETTINGS_PATH);`
 - All the exception handling would occur in the BaseExtension

### What alternatives have you considered?

Leaving it as is so extension authors can handle it themselves.

### Do you have any additional context?

DRY

### PR DESCRIPTION
**Pull request by mloufra**
Signed-off-by: mloufra <mloufra@amazon.com>

### Description
Move the `extensionSettings` field and the initialization in the constructor and all the exception handling in `HelloWorldExtension` to `BaseExtension`.

### Issues Resolved
Fixes #182

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:41]: This path constant (if used) needs to stay in the Extension subclass/implementation.  This `BaseExtension` class shouldn't have any extension-specific code.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:51]: The constructor here should take an argument that lets this class find the `ExtensionSettings`. This is how the subclass will pass its settings in.  You will need one or two constructors:
1. Definitely include one that allows the user to just instantiate extension settings without file reading, e.g., the constructor in the subclass would call `super(new ExtensionSettings("hello-world", "127.0.0.1", "4532"));`
2. Allow the YAML path one of these two ways:
   - Create a `String` constructor and just pass in the path and do the file reading here
   - Pass the resolved path to the constructor you created in bullet 1, like `super(Extension.readSettingsFromYaml(EXTENSION_SETTINGS_PATH));`.  This keeps it to only one constructor, but also introduces complexity for the user.  So I think I'd prefer just passing the String.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:67]: You can't inherit `static` methods so this should definitely not be `protected`. But I think this code is going to be moved into the constructor anyway (along with the javadoc outlining the two options).

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:46]: You'll be setting this in the constructor, so make it `private final`.  And add a getter (or move it up from the subclass) that overrides `getExtensionSettings()`.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:41]: We are only ever setting this once, in the constructor, so it should be `final`.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:46]: I don't think we need this constructor any more now that we have others, and we want to require subclasses to provide settings.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:54]: This does not throw `IOException` as it is caught, and an `ExceptionInInitializerError` is thrown instead.  That doesn't need to be added to a throws block as it's a runtime exception.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtension.java:64]: Please also add a constructor that takes an `ExtensionSettings` object and sets it directly.

Optionally you can change the `this.settings = ` in the other constructor to call the new one you created like `this(Extension.readSettingsFromYaml(path));`.

File: src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java:44]: Can we keep this as the first line of the constructor javadoc?

File: src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java:44]: Don't think we need this `@throws`.

File: src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java:46]: Don't think we need this `@throws`.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Looking good!  A few more tweaks needed!

**Comment by mloufra** [comment]
> Looking good! A few more tweaks needed!

Thank you for you comment @dbwiddis

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (mloufra) [reply src/main/java/org/opensearch/sdk/BaseExtension.java:64]: @dbwiddis I add new constructor in `BaseExtension` which takes ExtensionSetting object and sets it directly. 
`protected BaseExtension(ExtensionSettings settings) {
        this.settings = settings;
    }`
Since the test class `ExtensionsRunnerForTest()` was using this empty constructor from `BaseExtension` in its constructor,  the `ExtensionsRunnerForTest()` Constructor need to add parameter. No sure we really need to pass the variable to the Constructor in `ExtensionsRunnerForTest()`.

**Review by owaiskazi19 — DISMISSED** [review]

**Review by dbwiddis — DISMISSED** [review]
LGTM!  Great job, I think this will really simplify the experience for users!

@owaiskazi19 FYI, once this is merged you may want to simplify the AD Extension constructor as well.

**Comment by owaiskazi19** [comment]
@dbwiddis can you resolve the comments if they look good to you? Also, let's hold merging this until we fix the build error

**Comment by dbwiddis** [comment]
> @dbwiddis can you resolve the comments if they look good to you? Also, let's hold merging this until we fix the build error

Done.

File: src/main/java/org/opensearch/sdk/BaseExtension.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/BaseExtension.java]: > @dbwiddis Why do we need add `final` for `ExtensionSettings settings`?

It's not so much a "need" as good programming practice if we know a value will never change.  It helps you:
1. Make sure if you add a new constructor that you set the value correctly
2. Avoid adding other code elsewhere that changes the value at a point later than instantiation (which could impact thread safety)

Adding `final` makes it a lot easier to not have to carefully review other accesses of the variable.

**Review by dbwiddis — APPROVED** [review]

**Review by owaiskazi19 — APPROVED** [review]

**Comment by owaiskazi19** [comment]
@dbwiddis I'll let you merge this one based on the sequence to avoid merge conflicts

**Pull request merged**
